HIP/hip-980.md

```diff
@@ -0,0 +1,120 @@
+for the simulations.
+
+## User stories
+
+- As a user of mirror node, I want to be able to query for the historical state of accounts, tokens, and other network entities.
+- As a user of mirror node, I want to be able to run locally an `eth_call`, `eth_estimateGas` or `eth_debugTraceTransaction` call that fails on a production environment against the corresponding public state.
+- As a mirror node developer, I want to be able to test new features locally against production state, so that I have better understanding of the feature behavior.
+
+## Specification
+
+The proposed enhancement involves adding a query parameter `timestamp` to the existing routes:
+
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/tokens` to return correct historical information about an account and their associated fungible tokens for a specific timestamp
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/nfts` to return correct historical information about an account and their associated nfts for a specific timestamp
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/crypto` to return correct historical information about an account's crypto allowances for a specific timestamp
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/nfts` to return correct historical information about an account's NFT allowances for a specific timestamp
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/tokens` to return correct historical information about an account's token allowances for a specific timestamp
+- `/api/v1/tokens/{tokenId}/nfts/{serialNumber}` to return correct historical information about an NFT for a specific timestamp
+
+**Timestamp Query Parameter**
+
+A new `timestamp` query parameter will be added to search for different resources. This query parameter will represent a consensus timestamp in a Unix format in `seconds.nanoseconds` format and will have an optional comparison operator. That is, it will search for results that `equal`, `greater` or `less` than the timestamp itself or its seconds component.
+
+The exact names of the operators will be `eq`, `gt`, `gte`, `lt` and `lte`. The user doesn't need to know the exact timestamp at which an entity was created or updated for it to function. The API will return the results that match the timestamp or the closest one to it.
+For example, if we have an `asc` ordering of the results, then `?timestamp=X`, `timestamp=eq:X`, or `timestamp=lte:X` will have the same behaviour and vice versa, where `X` is the desired timestamp.
+
+```
+/api/v1/accounts/{idOrAliasOrEvmAddress}/tokens?timestamp={value}
+/api/v1/accounts/{idOrAliasOrEvmAddress}/nfts?timestamp={value}
+/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/crypto?timestamp={value}
+/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/nfts?timestamp={value}
+/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/tokens?timestamp={value}
+/api/v1/tokens/{tokenId}/nfts/{serialNumber}?timestamp={value}
```

Comment:
Missing contracts 
```suggestion
/api/v1/contracts/{idOrAliasOrEvmAddress}?timestamp={value}
/api/v1/tokens/{tokenId}/nfts/{serialNumber}?timestamp={value}
```

Reply:
The contracts API already utilizes a timestamp filter.

Reply:
Gotcha